compiler/rustc_lint/src/let_underscore.rs

```diff
@@ -0,0 +1,166 @@
+use crate::{LateContext, LateLintPass, LintContext};
+use rustc_errors::Applicability;
+use rustc_hir as hir;
+use rustc_middle::{lint::LintDiagnosticBuilder, ty};
+use rustc_span::Symbol;
+
+declare_lint! {
+    /// The `let_underscore_drop` lint checks for statements which don't bind
+    /// an expression which has a non-trivial Drop implementation to anything,
+    /// causing the expression to be dropped immediately instead of at end of
+    /// scope.
+    ///
+    /// ### Example
+    /// ```
+    /// struct SomeStruct;
+    /// impl Drop for SomeStruct {
+    ///     fn drop(&mut self) {
+    ///         println!("Dropping SomeStruct");
+    ///     }
+    /// }
+    ///
+    /// fn main() {
+    ///    #[warn(let_underscore_drop)]
+    ///     // SomeStuct is dropped immediately instead of at end of scope,
+    ///     // so "Dropping SomeStruct" is printed before "end of main".
+    ///     // The order of prints would be reversed if SomeStruct was bound to
+    ///     // a name (such as "_foo").
+    ///     let _ = SomeStruct;
+    ///     println!("end of main");
+    /// }
+    /// ```
+    ///
+    /// {{produces}}
+    ///
+    /// ### Explanation
+    ///
+    /// Statements which assign an expression to an underscore causes the
+    /// expression to immediately drop instead of extending the expression's
+    /// lifetime to the end of the scope. This is usually unintended,
+    /// especially for types like `MutexGuard`, which are typically used to
+    /// lock a mutex for the duration of an entire scope.
+    ///
+    /// If you want to extend the expression's lifetime to the end of the scope,
+    /// assign an underscore-prefixed name (such as `_foo`) to the expression.
+    /// If you do actually want to drop the expression immediately, then
+    /// calling `std::mem::drop` on the expression is clearer and helps convey
+    /// intent.
+    pub LET_UNDERSCORE_DROP,
+    Allow,
```

Comment:
Ah! Can you push a new commit with this being `Deny`? That way we can do a crater run to see how bad this heuristic would actually be in the wild. Depending on the results of that run, we'd make this either `Warn` or `Allow`, as you have here.

@joshtriplett this is necessary to catch *all* cases of the RAII pattern, but we need to verify what the S/N ratio is for the simplistic heuristic. Otherwise we'd need to extend `LET_UNDERSCORE_LOCK` to also check for the mutexes in parkinglot and tokio, at least.

Reply:
Done! I will note that this ends up triggering on code in the compiler itself, which I'm guessing means it's going to be extremely noisy.

Reply:
@estebank I would hypothesize that `let _ = returns_type_implementing_drop()` is quite common, too much so to turn on by default.

Reply:
Case in point:

```
   Compiling rustc-demangle v0.1.21
error: non-binding let on a type that implements `Drop`
   --> library/alloc/src/vec/into_iter.rs:324:21
    |
324 |                     let _ = RawVec::from_raw_parts_in(self.0.buf.as_ptr(), self.0.cap, alloc);
    |
    |
    = note: `#[deny(let_underscore_drop)]` on by default
help: consider binding to an unused variable to avoid immediately dropping the value
    |
324 |                     let _unused = RawVec::from_raw_parts_in(self.0.buf.as_ptr(), self.0.cap, alloc);
help: consider immediately dropping the value
    |
    |
324 |                     drop(RawVec::from_raw_parts_in(self.0.buf.as_ptr(), self.0.cap, alloc));

error: could not compile `alloc` due to previous error
warning: build failed, waiting for other jobs to finish...
Build completed unsuccessfully in 0:04:32
```

:-/

Reply:
@a2aaron this is a problem for what I wanted to do. Did you see how many cases are in the rustc repo? Because if `rustc` doesn't build, then we can't run crater to collect numbers on how many crates would suddenly start warning on this. I guess path inspecting against a denylist of common crates (or better, stabilizing an attribute that they can annotate with that `rustc` checks for) might be warranted in the end :-/

Reply:
@estebank There's a lot. Setting `let_underscore_drop` to `Warn` and building with `--warnings warn` turned on, I got 28 warnings. Most of these were in `std`. It seems like 18 of them are caused by one macro (`rtprintpanic!`).

Reply:
That's actually not as many as I thought there would be! 😅

Reply:
Should I try manually fixing up all of those cases?

Reply:
Only if you have time, otherwise we can defer on that and focus on the specific denylist of std mutexes.

Reply:
I think I'll defer on it--the lint is too noisy and annoying to fix up right now, and every instance of it firing has been on types where it really doesn't matter that it gets dropped earlier (it's firing on a lot of `Box`s or various `Result` types). I've pushed a commit making `let_underscore_drop` allow by default for now.